crates/cext/src/circuit.rs

```diff
@@ -409,6 +413,115 @@ pub struct OpCounts {
     len: usize,
 }
 
+#[inline]
+fn conjugate(matrix: &ArrayView2<Complex64>) -> Array2<Complex64> {
+    Array2::from_shape_fn((matrix.nrows(), matrix.ncols()), |(i, j)| {
+        matrix[(j, i)].conj()
+    })
+}
+
+/// Check an [ArrayType] represents a unitary matrix. Uses an element-wise check; if
+/// any element in ``conjugate(matrix) * matrix`` differs from the identity by more than ``tol``
+/// (in magnitude), the matrix is not considered unitary.
+fn is_unitary(matrix: &ArrayType, tol: f64) -> bool {
+    let not_unitary = match matrix {
+        ArrayType::OneQ(mat) => (mat.adjoint() * mat - Matrix2::identity())
+            .iter()
+            .any(|val| val.abs() > tol),
+        ArrayType::TwoQ(mat) => (mat.adjoint() * mat - Matrix4::identity())
+            .iter()
+            .any(|val| val.abs() > tol),
+        ArrayType::NDArray(mat) => {
+            let product = mat.dot(&conjugate(&mat.view()));
+            let n = mat.nrows();
+            let one = Complex64::new(1.0, 0.0);
+            product.iter().enumerate().any(|(index, value)| {
```

Comment:
There is https://docs.rs/ndarray/latest/ndarray/struct.ArrayBase.html#method.indexed_iter which does this natively in a single iterator method.